### DISCUSSION TITLE
Best way to tell eglot to look for my python language servers in anaconda?

### DISCUSSION THREAD

**Discussion by qazmonk** [joaotavora/eglot#972] [Q&A]
I generally use anaconda to manage my python packages on my local and remote machines, including language servers. When I start a new emacs instance and open a python file I usually get an eglot error: `(error "None of ’pylsp, pyls, pyright-langserver’ are valid executables")` because I haven't yet activated my anaconda base env with conda.el. This is a small problem with a variety of fixes but I'm wondering what other people do or if I'm missing something obvious. It seems like there are several options:

1. Put `(conda-env-activate "base")` in my init file.
    Seems like asking for trouble when I inevitably run emacs  on a machine without anaconda, also will slow down my init time.
2. Turn on `conda-env-autoactivate-mode`.
    I was doing this but I kept getting error messages from conda.el when switching to buffers with python mode but no underlying file, or when editing remote python files.
3. Soft link my conda installed language server to `/usr/local/bin`.
    Could be the most robust option, but I'm not familiar enough with conda internals to know if this is bad for some reason.
4. Add `anaconda/bin` to my path.
    This feels like what should be happening, conda init is in my .zshrc but when I launch emacs from the OSX launcher it doesn't seem to activate.

Am I missing something? Do you do one of these options?

**Comment by changephilip** [comment]
My workflow is as following:
1.  in (base) environment `pip3 install python-lsp-server` 
2. `M-x pyvenv-activate` to select `~/miniconda3/envs/abc `
3. then `M-x eglot-reconnect`.

**Reply by changephilip** [reply]
if you edit file through tramp ,then install python-lsp-server in each conda env and 
remeber to comment this line in init.el
`;;(add-to-list 'tramp-remote-path 'tramp-own-remote-path)`
after open remote file with tramp , `M-x eglot` to select your `remote:~/miniconda3/envs/abc/bin/pylsp`

**Reply by QiangF** [reply]
I think you don't need to install pylsp in each conda env. I use archlinux, I install python-lsp-server with pacman system wide, and specify conda-project-env-path in .dir-locals.el, works very well.

**Comment by qianjiaqiang** [comment]
QiangF's method works well for me now.
Create .dir-locals.el in the remote project directory:
```
((nil
  . ((eglot-workspace-configuration
      . (:pylsp (:plugins (:jedi
                           (:environment  "~/miniconda3/envs/abc"))))))))

```
Uncomment this line to help emacs find remote pylsp in` ~/miniconda3/bin/`
`(add-to-list 'tramp-remote-path 'tramp-own-remote-path)
`
and add this line:
`(setq enable-remote-dir-locals t)`

> I think you don't need to install pylsp in each conda env. I use archlinux, I install python-lsp-server with pacman system wide, and specify conda-project-env-path in .dir-locals.el, works very well.

**Reply by t1anchen** [reply]
it would be better to `pyvenv-activate` the `~/miniconda3/envs/abc` first to avoid contaminating global environment, e.g.

```elisp
;;; Directory Local Variables
;;; For more information see (info "(emacs) Directory Variables")

((nil . ((eval . (let ((venv-dir "/path/to/conda/env/"))
                   (pyvenv-activate venv-dir)
                   (setq-default
                    eglot-workspace-configuration
                    (lambda (&rest args)
                      `((:pylsp .
                                (:plugins
                                 (:jedi_completion
                                  (:include_params t :fuzzy t)
                                  :jedi (:environment ,venv-dir)
                                  :black (:enabled t)
                                  ))))))
                   (add-hook 'projectile-after-switch-project-hook #'(lambda () (pyvenv-deactivate))))))))
```
